Inefficient compression for short/unaligned writes

Calls to Write should be buffered to avoid emitting blocks that are shorter than necessary (and by extension inefficiently compressed). I put together a worst-case albeit completely unrealistic example demonstrating the inefficiency. 

https://github.com/bmatsuo/go-snappystream/blob/inefficient-short-writes/writer_test.go

```
--- FAIL: TestWriter_shortWriteEfficiency (0.00 seconds)
    writer_test.go:29: batch write size 27 not equal short write size 1110 (x41.1)
FAIL
exit status 1
FAIL    github.com/mreiferson/go-snappystream   0.010s
```

While the test case will not be seen in practice short writes will occur when writes are not made in a multiple of MaxBlockSize bytes.

IMO, simply wrapping Writer in a bufio.Writer is not adequate in the general case. I don't see a particular benefit to that over internal buffering. And chunks which cannot be compressed, or application specific logic that flushes the buffer prematurely will cause short writes. (if that is the intended solution it should be documented)

I think an API similar to "compress/gzip" should be exported. That is, there should be a type that implements

```
interface {
    io.WriteCloser
    Flush() error
}
```

--- a/readwrite_test.go
+++ b/readwrite_test.go
@@ -10,15 +10,25 @@ import (
 
 const TestFileSize = 10 << 20 // 10MB
 
+// dummyBytesReader returns an io.Reader that avoids buffering optimizations
+// in io.Copy. This can be considered a 'worst-case' io.Reader as far as writer
+// frame alignment goes.
+//
+// Note: io.Copy uses a 32KB buffer internally as of Go 1.3, but that isn't
+// part of its public API (undocumented).
+func dummyBytesReader(p []byte) io.Reader {
+	return ioutil.NopCloser(bytes.NewReader(p))
+}
+
 func testWriteThenRead(t *testing.T, name string, bs []byte) {
 	var buf bytes.Buffer
 	w := NewWriter(&buf)
-	n, err := w.Write(bs)
+	n, err := io.Copy(w, dummyBytesReader(bs))
 	if err != nil {
 		t.Errorf("write %v: %v", name, err)
 		return
 	}
-	if n != len(bs) {
+	if n != int64(len(bs)) {
 		t.Errorf("write %v: wrote %d bytes (!= %d)", name, n, len(bs))
 		return
 	}
@@ -31,8 +41,8 @@ func testWriteThenRead(t *testing.T, name string, bs []byte) {
 		t.Errorf("read %v: %v", name, err)
 		return
 	}
-	n = len(gotbs)
-	if n != len(bs) {
+	n = int64(len(gotbs))
+	if n != int64(len(bs)) {
 		t.Errorf("read %v: read %d bytes (!= %d)", name, n, len(bs))
 		return
 	}
@@ -43,7 +53,48 @@ func testWriteThenRead(t *testing.T, name string, bs []byte) {
 	}
 
 	c := float64(len(bs)) / float64(enclen)
-	t.Logf("%v compression ratio %.03g", name, c)
+	t.Logf("%v compression ratio %.03g (%d byte reduction)", name, c, len(bs)-enclen)
+}
+
+func testBufferedWriteThenRead(t *testing.T, name string, bs []byte) {
+	var buf bytes.Buffer
+	w := NewBufferedWriter(&buf)
+	n, err := io.Copy(w, dummyBytesReader(bs))
+	if err != nil {
+		t.Errorf("write %v: %v", name, err)
+		return
+	}
+	if n != int64(len(bs)) {
+		t.Errorf("write %v: wrote %d bytes (!= %d)", name, n, len(bs))
+		return
+	}
+	err = w.Close()
+	if err != nil {
+		t.Errorf("close %v: %v", name, err)
+		return
+	}
+
+	enclen := buf.Len()
+
+	r := NewReader(&buf, true)
+	gotbs, err := ioutil.ReadAll(r)
+	if err != nil {
+		t.Errorf("read %v: %v", name, err)
+		return
+	}
+	n = int64(len(gotbs))
+	if n != int64(len(bs)) {
+		t.Errorf("read %v: read %d bytes (!= %d)", name, n, len(bs))
+		return
+	}
+
+	if !bytes.Equal(gotbs, bs) {
+		t.Errorf("%v: unequal decompressed content", name)
+		return
+	}
+
+	c := float64(len(bs)) / float64(enclen)
+	t.Logf("%v compression ratio %.03g (%d byte reduction)", name, c, len(bs)-enclen)
 }
 
 func TestWriterReader(t *testing.T) {
@@ -59,6 +110,22 @@ func TestWriterReader(t *testing.T) {
 		t.Fatal(err)
 	}
 	testWriteThenRead(t, "random", p)
+
+}
+
+func TestBufferedWriterReader(t *testing.T) {
+	testBufferedWriteThenRead(t, "simple", []byte("test"))
+	testBufferedWriteThenRead(t, "manpage", testDataMan)
+	testBufferedWriteThenRead(t, "json", testDataJSON)
+
+	p := make([]byte, TestFileSize)
+	testBufferedWriteThenRead(t, "constant", p)
+
+	_, err := rand.Read(p)
+	if err != nil {
+		t.Fatal(err)
+	}
+	testBufferedWriteThenRead(t, "random", p)
 
 }
 
@@ -96,8 +163,16 @@ func BenchmarkWriterManpage(b *testing.B) {
 	benchmarkWriterBytes(b, testDataMan)
 }
 
+func BenchmarkBufferedWriterManpage(b *testing.B) {
+	benchmarkBufferedWriterBytes(b, testDataMan)
+}
+
 func BenchmarkWriterJSON(b *testing.B) {
 	benchmarkWriterBytes(b, testDataJSON)
+}
+
+func BenchmarkBufferedWriterJSON(b *testing.B) {
+	benchmarkBufferedWriterBytes(b, testDataJSON)
 }
 
 // BenchmarkWriterRandom tests basically uncompressable data.
@@ -111,6 +186,16 @@ func BenchmarkWriterRandom(b *testing.B) {
 	benchmarkWriterBytes(b, randp)
 }
 
+func BenchmarkBufferedWriterRandom(b *testing.B) {
+	size := TestFileSize
+	randp := make([]byte, size)
+	_, err := rand.Read(randp)
+	if err != nil {
+		b.Fatal(err)
+	}
+	benchmarkBufferedWriterBytes(b, randp)
+}
+
 // BenchmarkWriterConstant tests maximally compressible data
 func BenchmarkWriterConstant(b *testing.B) {
 	size := TestFileSize
@@ -118,17 +203,43 @@ func BenchmarkWriterConstant(b *testing.B) {
 	benchmarkWriterBytes(b, zerop)
 }
 
+func BenchmarkBufferedWriterConstant(b *testing.B) {
+	size := TestFileSize
+	zerop := make([]byte, size)
+	benchmarkBufferedWriterBytes(b, zerop)
+}
+
 func benchmarkWriterBytes(b *testing.B, p []byte) {
 	b.SetBytes(int64(len(p)))
-	w := NewWriter(ioutil.Discard)
 	b.StartTimer()
 	for i := 0; i < b.N; i++ {
-		n, err := w.Write(p)
+		w := NewWriter(ioutil.Discard) // create every time for stream identifier
+		n, err := io.Copy(w, dummyBytesReader(p))
 		if err != nil {
 			b.Fatalf(err.Error())
 		}
-		if n != len(p) {
+		if n != int64(len(p)) {
 			b.Fatalf("wrote wrong amount %d != %d", n, len(p))
+		}
+	}
+	b.StopTimer()
+}
+
+func benchmarkBufferedWriterBytes(b *testing.B, p []byte) {
+	b.SetBytes(int64(len(p)))
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		w := NewBufferedWriter(ioutil.Discard)
+		n, err := io.Copy(w, dummyBytesReader(p))
+		if err != nil {
+			b.Fatalf(err.Error())
+		}
+		if n != int64(len(p)) {
+			b.Fatalf("wrote wrong amount %d != %d", n, len(p))
+		}
+		err = w.Close()
+		if err != nil {
+			b.Fatal(err)
 		}
 	}
 	b.StopTimer()
@@ -138,8 +249,16 @@ func BenchmarkReaderManpage(b *testing.B) {
 	benchmarkReaderDiscard(b, testDataMan)
 }
 
+func BenchmarkReaderManpage_buffered(b *testing.B) {
+	benchmarkReaderDiscard_buffered(b, testDataMan)
+}
+
 func BenchmarkReaderJSON(b *testing.B) {
 	benchmarkReaderDiscard(b, testDataJSON)
+}
+
+func BenchmarkReaderJSON_buffered(b *testing.B) {
+	benchmarkReaderDiscard_buffered(b, testDataJSON)
 }
 
 // BenchmarkReaderRandom tests basically uncompressable data.
@@ -153,6 +272,16 @@ func BenchmarkReaderRandom(b *testing.B) {
 	benchmarkReaderDiscard(b, randp)
 }
 
+func BenchmarkReaderRandom_buffered(b *testing.B) {
+	size := TestFileSize
+	randp := make([]byte, size)
+	_, err := rand.Read(randp)
+	if err != nil {
+		b.Fatal(err)
+	}
+	benchmarkReaderDiscard_buffered(b, randp)
+}
+
 // BenchmarkReaderConstant tests maximally compressible data
 func BenchmarkReaderConstant(b *testing.B) {
 	size := TestFileSize
@@ -160,10 +289,16 @@ func BenchmarkReaderConstant(b *testing.B) {
 	benchmarkReaderDiscard(b, zerop)
 }
 
+func BenchmarkReaderConstant_buffered(b *testing.B) {
+	size := TestFileSize
+	zerop := make([]byte, size)
+	benchmarkReaderDiscard_buffered(b, zerop)
+}
+
 func benchmarkReaderDiscard(b *testing.B, p []byte) {
 	var buf bytes.Buffer
 	w := NewWriter(&buf)
-	_, err := w.Write(p)
+	_, err := io.Copy(w, dummyBytesReader(p))
 	if err != nil {
 		b.Fatal("pre-test compression: %v", err)
 	}
@@ -172,7 +307,35 @@ func benchmarkReaderDiscard(b *testing.B, p []byte) {
 	b.SetBytes(int64(len(encp)))
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		r := NewReader(bytes.NewReader(encp), true)
+		r := NewReader(dummyBytesReader(encp), true)
+		n, err := io.Copy(ioutil.Discard, r)
+		if err != nil {
+			b.Fatalf(err.Error())
+		}
+		if n != int64(len(p)) {
+			b.Fatalf("read wrong amount %d != %d", n, len(p))
+		}
+	}
+	b.StopTimer()
+}
+
+func benchmarkReaderDiscard_buffered(b *testing.B, p []byte) {
+	var buf bytes.Buffer
+	w := NewBufferedWriter(&buf)
+	_, err := io.Copy(w, dummyBytesReader(p))
+	if err != nil {
+		b.Fatal("pre-test compression: %v", err)
+	}
+	err = w.Close()
+	if err != nil {
+		b.Fatal("pre-test compression: %v", err)
+	}
+	encp := buf.Bytes()
+
+	b.SetBytes(int64(len(encp)))
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		r := NewReader(dummyBytesReader(encp), true)
 		n, err := io.Copy(ioutil.Discard, r)
 		if err != nil {
 			b.Fatalf(err.Error())

--- a/writer.go
+++ b/writer.go
@@ -1,6 +1,7 @@
 package snappystream
 
 import (
+	"bufio"
 	"errors"
 	"fmt"
 	"hash/crc32"
@@ -8,6 +9,83 @@ import (
 
 	"code.google.com/p/snappy-go/snappy"
 )
+
+var errClosed = fmt.Errorf("closed")
+
+// BufferedWriter is an io.WriteCloser with behavior similar to writers
+// returned by NewWriter but it buffers written data, maximizing block size (to
+// improve the output compression ratio) at the cost of speed. Benefits over
+// NewWriter are most noticible when individual writes are small and when
+// streams are long.
+//
+// Failure to call a BufferedWriter's Close or Flush methods after it is done
+// being written to will likely result in missing data frames which will be
+// undetectable in the decoding process.
+type BufferedWriter struct {
+	err error
+	w   *writer
+	bw  *bufio.Writer
+}
+
+// NewBufferedWriter allocates and returns a BufferedWriter with an internal
+// buffer of MaxBlockSize bytes.  If an error occurs writing a block to w, all
+// future writes will fail with the same error.  After all data has been
+// written, the client should call the Flush method to guarantee all data has
+// been forwarded to the underlying io.Writer.
+func NewBufferedWriter(w io.Writer) *BufferedWriter {
+	_w := NewWriter(w).(*writer)
+	return &BufferedWriter{
+		w:  _w,
+		bw: bufio.NewWriterSize(_w, MaxBlockSize),
+	}
+}
+
+// Write buffers p internally, encoding and writing a block to the underlying
+// buffer if the buffer grows beyond MaxBlockSize bytes.  The returned int
+// will be 0 if there was an error and len(p) otherwise.
+func (w *BufferedWriter) Write(p []byte) (int, error) {
+	if w.err != nil {
+		return 0, w.err
+	}
+
+	_, w.err = w.bw.Write(p)
+	if w.err != nil {
+		return 0, w.err
+	}
+
+	return len(p), nil
+}
+
+// Flush encodes and writes a block with the contents of w's internal buffer to
+// the underlying writer even if the buffer does not contain a full block of
+// data (MaxBlockSize bytes).
+func (w *BufferedWriter) Flush() error {
+	if w.err == nil {
+		w.err = w.bw.Flush()
+	}
+
+	return w.err
+}
+
+// Close flushes w's internal buffer and tears down internal data structures.
+// After a successful call to Close method calls on w return an error.  Close
+// makes no attempt to close the underlying writer.
+func (w *BufferedWriter) Close() error {
+	if w.err != nil {
+		return w.err
+	}
+
+	w.err = w.bw.Flush()
+	w.w = nil
+	w.bw = nil
+
+	if w.err != nil {
+		return w.err
+	}
+
+	w.err = errClosed
+	return nil
+}
 
 type writer struct {
 	writer io.Writer
@@ -18,21 +96,17 @@ type writer struct {
 	sentStreamID bool
 }
 
-// NewWriter returns an io.Writer interface to the snappy framed stream format.
+// NewWriter returns an io.Writer that writes its input to an underlying
+// io.Writer encoded as a snappy framed stream.  A stream identifier block is
+// written to w preceding the first data block.  The returned writer will never
+// emit a block with length in bytes greater than MaxBlockSize+4 nor one
+// containing more than MaxBlockSize bytes of (uncompressed) data.
 //
-// It transparently handles sending the stream identifier, calculating
-// checksums, and compressing/framing blocks.
-//
-// Internally, a buffer is maintained to hold a compressed
-// block.  It will automatically re-size up the the largest
-// block size, 65536.
-//
-// For each Write, the returned length will only ever be len(p) or 0, regardless
-// of the length of *compressed* bytes written to the wrapped io.Writer.
-//
-// If the returned length is 0 then error will be non-nil.
-//
-// If len(p) exceeds 65536, the slice will be automatically chunked into smaller blocks.
+// For each Write, the returned length will only ever be len(p) or 0,
+// regardless of the length of *compressed* bytes written to the wrapped
+// io.Writer.  If the returned length is 0 then error will be non-nil.  If
+// len(p) exceeds 65536, the slice will be automatically chunked into smaller
+// blocks which are all emitted before the call returns.
 func NewWriter(w io.Writer) io.Writer {
 	return &writer{
 		writer: w,

--- a/writer_test.go
+++ b/writer_test.go
@@ -1,0 +1,126 @@
+package snappystream
+
+import (
+	"bytes"
+	"io/ioutil"
+	"log"
+	"testing"
+)
+
+// This test ensures that all BufferedWriter methods fail after Close has been
+// called.
+func TestBufferedWriterClose(t *testing.T) {
+	w := NewBufferedWriter(ioutil.Discard)
+	err := w.Close()
+	if err != nil {
+		log.Fatalf("closing empty BufferedWriter: %v", err)
+	}
+	err = w.Close()
+	if err == nil {
+		log.Fatalf("successful close after close")
+	}
+	err = w.Flush()
+	if err == nil {
+		log.Fatalf("successful flush after close")
+	}
+	_, err = w.Write([]byte("abc"))
+	if err == nil {
+		log.Fatalf("successful write after close")
+	}
+}
+
+// This test simply checks that buffering has an effect in a situation where it
+// is know it should.
+func TestBufferedWriter_compression(t *testing.T) {
+	p := []byte("hello snappystream!")
+	n := 10
+
+	var shortbuf bytes.Buffer
+	w := NewWriter(&shortbuf)
+	for i := 0; i < n; i++ {
+		n, err := w.Write(p)
+		if err != nil {
+			t.Fatalf("writer error: %v", err)
+		}
+		if n != len(p) {
+			t.Fatalf("short write: %d", n)
+		}
+	}
+
+	var buf bytes.Buffer
+	bw := NewBufferedWriter(&buf)
+	for i := 0; i < n; i++ {
+		n, err := bw.Write(p)
+		if err != nil {
+			t.Fatalf("buffered writer error: %v", err)
+		}
+		if n != len(p) {
+			t.Fatalf("short write: %d", n)
+		}
+	}
+	err := bw.Close()
+	if err != nil {
+		t.Fatalf("closing buffer: %v", err)
+	}
+
+	uncompressed := int64(n) * int64(len(p))
+	compressed := shortbuf.Len()
+	bufcompressed := buf.Len()
+
+	if compressed <= bufcompressed {
+		t.Fatalf("no benefit from buffering (%d <= %d)", shortbuf.Len(), buf.Len())
+	}
+
+	c := float64(uncompressed) / float64(compressed)
+	bufc := float64(uncompressed) / float64(bufcompressed)
+	improved := bufc / c
+
+	t.Logf("BufferedWriter compression ratio %g (%.03g factor improvement over %g)", bufc, improved, c)
+}
+
+// This tests ensures flushing after every write is equivalent to using
+// NewWriter directly.
+func TestBufferedWriterFlush(t *testing.T) {
+	p := []byte("hello snappystream!")
+	n := 10
+
+	var shortbuf bytes.Buffer
+	w := NewWriter(&shortbuf)
+	for i := 0; i < n; i++ {
+		n, err := w.Write(p)
+		if err != nil {
+			t.Fatalf("writer error: %v", err)
+		}
+		if n != len(p) {
+			t.Fatalf("short write: %d", n)
+		}
+	}
+
+	var buf bytes.Buffer
+	bw := NewBufferedWriter(&buf)
+	for i := 0; i < n; i++ {
+		n, err := bw.Write(p)
+		if err != nil {
+			t.Fatalf("buffered writer error: %v", err)
+		}
+		if n != len(p) {
+			t.Fatalf("short write: %d", n)
+		}
+		err = bw.Flush()
+		if err != nil {
+			t.Fatal("flush: %v", err)
+		}
+	}
+	err := bw.Close()
+	if err != nil {
+		t.Fatalf("closing buffer: %v", err)
+	}
+
+	if shortbuf.Len() != buf.Len() {
+		t.Fatalf("unexpected size: %d != %d", shortbuf.Len(), buf.Len())
+	}
+
+	if !bytes.Equal(shortbuf.Bytes(), buf.Bytes()) {
+		t.Fatalf("unexpected bytes")
+	}
+}

